src/expr/src/linear.rs

```diff
@@ -533,4 +546,51 @@ impl MapFilterProject {
         // This should probably resemble existing scalar cse.
         unimplemented!()
     }
+
+    /// Removes unused expressions from `self.expressions`.
+    ///
+    /// Expressions are "used" if they are relied upon by any output columns
+    /// or any predicates, even transitively. Any expressions that are not
+    /// relied upon in this way can be discarded.
+    pub fn remove_undemanded(&mut self) {
+        // Determine the demanded expressions to remove irrelevant ones.
+        let mut demand = std::collections::HashSet::new();
+        for (_index, pred) in self.predicates.iter() {
+            demand.extend(pred.support());
+        }
+        demand.extend(self.projection.iter().cloned());
+        for index in (0..self.expressions.len()).rev() {
+            if demand.contains(&(self.input_arity + index)) {
+                demand.extend(self.expressions[index].support());
+            }
+        }
+
+        // Expressions that are not demanded can be removed, and all
+        // remaining indexes updated.
+        let mut remap = HashMap::new();
+        for index in 0..self.input_arity {
+            remap.insert(index, index);
+        }
+        // Retain demanded expressions, and record their new locations.
+        let mut index = self.input_arity;
+        self.expressions.retain(|_expr| {
```

Comment:
minor: this closure for `retain` was challenging to understand. I can't tell if this code is perf sensitive but if not, could we build up a new array in a for loop and then put the new array in `self.expressions` instead. feel free to disregard

Reply:
I can't explain why it is so complicated. I don't believe it is performance sensitive, and should be rewritten for clarity. Brb.

Reply:
Or, hrm. Maybe I was scared of expression cloning, which can be arbitrarily more expensive than this. I'll try out a new version and we can look at it.

Reply:
How does this look:
```rust
        // Retain demanded expressions, and record their new locations.
        let mut new_expressions = Vec::new();
        for (index, expr) in self.expressions.drain(..).enumerate() {
            if demand.contains(&(index + self.input_arity)) {
                remap.insert(index + self.input_arity, remap.len());
                new_expressions.push(expr);
            }
        }
        self.expressions = new_expressions;
```

Reply:
yeah that looks a lot friendlier :)